biojava-core/src/main/java/org/biojava/nbio/core/search/io/Hsp.java

```diff
@@ -0,0 +1,229 @@
+package org.biojava.nbio.core.search.io;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import org.biojava.nbio.core.alignment.SimpleAlignedSequence;
+import org.biojava.nbio.core.alignment.SimpleSequencePair;
+import org.biojava.nbio.core.alignment.template.AlignedSequence.Step;
+import org.biojava.nbio.core.alignment.template.SequencePair;
+import org.biojava.nbio.core.exceptions.CompoundNotFoundException;
+import org.biojava.nbio.core.sequence.DNASequence;
+import org.biojava.nbio.core.sequence.ProteinSequence;
+import org.biojava.nbio.core.sequence.RNASequence;
+import org.biojava.nbio.core.sequence.compound.AminoAcidCompoundSet;
+import org.biojava.nbio.core.sequence.compound.DNACompoundSet;
+import org.biojava.nbio.core.sequence.template.Compound;
+import org.biojava.nbio.core.sequence.template.Sequence;
+
+/**
+ * This class models a search Hsp.
+ * You will retrieve a list of this using iterator of a Hit
+ * 
+ * Designed by Paolo Pavan.
+ * You may want to find my contacts on Github and LinkedIn for code info 
+ * or discuss major changes.
+ * https://github.com/paolopavan
+ * 
+ * @author Paolo Pavan
+ */
+
+public abstract class Hsp <S extends Sequence<C>, C extends Compound> {
+    private Integer hspNum;
+    private Double hspBitScore;
+    private Integer hspScore;
+    private Double hspEvalue;
+    private Integer hspQueryFrom;
+    private Integer hspQueryTo;
+    private Integer hspHitFrom;
+    private Integer hspHitTo;
+    private Integer hspQueryFrame;
+    private Integer hspHitFrame;
+    private Integer hspIdentity;
+    private Integer hspPositive;
+    private Integer hspGaps;
+    private Integer hspAlignLen;
+    private String hspQseq;
+    private String hspHseq;
+    private String hspIdentityString;
+    private Double percentageIdentity = null;
+    private Integer mismatchCount = null;
+    private SimpleSequencePair<S, C> returnAln;
+    
+    /**
+     * Experimental.
+     * Wants to return an hashcode designed to allow conceptual comparisons of search results.
+     * Fields unrelated to search are deliberately not considered.
+     * @return 
+     */
+    public int hashCode(){
+        String allInOne = hspQseq+"\n"+hspIdentityString+"\n"+hspHseq;
+        return allInOne.hashCode();
+    }
+    
+    @Override
+    public boolean equals(Object o){
+        if (!(o instanceof Hsp)) return false;
+        
+        return o.hashCode() == this.hashCode();
+    }

```

Comment:
I have changed the full stuff. I have revisited the way equality is
evaluated and I have modified properties accordingly.

2015-09-22 20:23 GMT+02:00 Jose Manuel Duarte <notifications@github.com>:

> In biojava-core/src/main/java/org/biojava/nbio/core/search/io/Hsp.java
> <https://github.com/biojava/biojava/pull/329#discussion_r40123811>:
>
> > +     * Experimental.
> > +     * Wants to return an hashcode designed to allow conceptual comparisons of search results.
> > +     * Fields unrelated to search are deliberately not considered.
> > +     * @return
> > +     */
> > +    public int hashCode(){
> > +        String allInOne = hspQseq+"\n"+hspIdentityString+"\n"+hspHseq;
> > +        return allInOne.hashCode();
> > +    }
> > +
> > +    @Override
> > +    public boolean equals(Object o){
> > +        if (!(o instanceof Hsp)) return false;
> > +
> > +        return o.hashCode() == this.hashCode();
> > +    }
>
> Two comments here:
>
>    -
>
>    equals should not be implemented by comparing hashCodes. Hash codes do
>    not guarantee equality. Hash equality is a necessary but not sufficient
>    condition for equality (2 different objects can have the same hash)
>    -
>
>    the best way to implement a hashCode is to follow the indications in this
>    SO thread
>    <http://stackoverflow.com/questions/113511/best-implementation-for-hashcode-method>.
>    Actually if you are using eclipse it can even implement such a thing for
>    you automatically, just go to 'Source' -> 'Generate hashCode and equals',
>    you then only need to select which fields of the class should be used (I
>    imagine other IDEs will have similar functions).
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/biojava/biojava/pull/329/files#r40123811>.
>